README.md

```diff
@@ -1,78 +1,406 @@
+
+Let's assume you only want to generate string values for the `firstName` field; this can be achieved by adding an `ofType` constraint for the field. With this constraint alone, the generator will output random strings containing characters from the unicode [Basic Multilingual Plane](https://en.wikipedia.org/wiki/Plane_(Unicode)#Basic_Multilingual_Plane). If you want a more restrictive character set, ths can be achieved by adding `matchingRegex`. 
+
+
+## Adding constraints
+
+Update the JSON profile as follows:
+
+~~~json
+{
+  "schemaVersion": "0.1",
+  "fields": [
+    { "name": "firstName" }
+  ],
+  "rules": [
+    {
+      "rule": "first name constraints",
+      "constraints": [
+        { "field": "firstName", "is": "ofType", "value": "string" },
+        { "field": "firstName", "is": "matchingRegex", "value": "[a-z]{1,10}" }
+      ]
+    }
+  ]
+}
+~~~
+
+Re-running generation now creates a file containing random strings that match the simple regex `[a-z]{1,10}`:
+
+~~~
+firstName
+"f"
+
+"draxonsa"
+"j"
+"srsub"
+
+"f"
+
+[...]
+~~~
+
+Fields are nullable by default, however, you can add a further constraint to a field to indicate it is 'not null'. If you add the following constraint, the generator will no longer output null values:
+
+~~~json
+"constraints": [
+  { "field": "firstName", "is": "ofType", "value": "string" },
+  { "field": "firstName", "is": "matchingRegex", "value": "[a-z]{1,10}" },
+  { "not": { "field": "firstName", "is": "null" } }
+]
+~~~
+
+The generator supports three different types of constraint. These are:
+
+ - **Predicates** - boolean-valued functions that define whether a given value as valid or invalid. In the above profile `ofType` and `matchingRegex` are examples of predicates.
+ - **Grammatical** - combine or modify other constraints. They are fully recursive; any grammatical constraint is a valid input to any other grammatical constraint. In the above profile `not` is an example of a grammatical constraint. 
+ - **Presentational** - provide additional formatting information, we'll cover these later.
+
+The current profile outputs random text strings for the `firstName` field. Depending on what you are intending to use this data for this may or may not be appropriate. For testing purposes, you are likely to want output data that has a lot of variability, however, if you are using the generator to create simulation data, you might want to use a regex that creates more realistic data, for example the regex `(Joh?n|Mar[yk])`.
+
+## Generating large datasets
+
+The generator has been designed to be fast and efficient, allowing you to generate large quantities of test and simulation data. If you supply a large number for the `--max-rows` option, the data will be streamed to the output file, with the progress / velocity reported during generation.
+
+```
+$ java -jar generator.jar generate --max-rows=10000 --overwrite profile.json out.csv
```

Comment:
is it better to keep the output filename the same, here its `out.csv` but earlier you use `output.csv`, minor point I know!

Reply:
good spot!